src/cargo/core/compiler/output_sbom.rs

```diff
@@ -0,0 +1,239 @@
+use serde::Serialize;
+
+use crate::{
+    core::{profiles::Profile, PackageId, Target, TargetKind},
+    util::Rustc,
+    CargoResult,
+};
+
+use super::{unit_graph::UnitDep, BuildRunner, CrateType, Unit};
+
+#[derive(Serialize, Clone, Debug, Copy)]
+#[serde(rename_all = "kebab-case")]
+enum SbomBuildType {
+    /// A package dependency
+    Normal,
+    /// A build script dependency
+    Build,
+}
+
+/// Typed version of a SBOM format version number.
+pub struct SbomFormatVersion<const V: u32>;
+
+impl<const V: u32> Serialize for SbomFormatVersion<V> {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        serializer.serialize_u32(V)
+    }
+}
+
+/// A package dependency
+#[derive(Serialize, Clone, Debug)]
+struct SbomDependency {
+    name: String,
+    package_id: PackageId,
+    version: String,
+    features: Vec<String>,
+}
+
+impl From<&UnitDep> for SbomDependency {
+    fn from(dep: &UnitDep) -> Self {
+        let package_id = dep.unit.pkg.package_id();
+        let name = package_id.name().to_string();
+        let version = package_id.version().to_string();
+        let features = dep
+            .unit
+            .features
+            .iter()
+            .map(|f| f.to_string())
+            .collect_vec();
+
+        Self {
+            name,
+            package_id,
+            version,
+            features,
+        }
+    }
+}
+
+#[derive(Serialize, Clone, Debug)]
+struct SbomPackage {
+    package_id: PackageId,
+    package: String,
+    version: String,
+    features: Vec<String>,
+    build_type: SbomBuildType,
+    extern_crate_name: String,
+    dependencies: Vec<SbomDependency>,
+}
+
+impl SbomPackage {
+    pub fn new(
+        dep: &UnitDep,
+        dependencies: Vec<SbomDependency>,
+        build_type: SbomBuildType,
+    ) -> Self {
+        let package_id = dep.unit.pkg.package_id();
+        let package = package_id.name().to_string();
+        let version = package_id.version().to_string();
+        let features = dep
+            .unit
+            .features
+            .iter()
+            .map(|f| f.to_string())
+            .collect_vec();
+
+        Self {
+            package_id,
+            package,
+            version,
+            features,
+            build_type,
+            extern_crate_name: dep.extern_crate_name.to_string(),
+            dependencies,
+        }
+    }
+}
+
+#[derive(Serialize)]
+struct SbomTarget {
+    kind: TargetKind,
+    crate_type: Option<CrateType>,
+    name: String,
+    edition: String,
+}
+
+impl From<&Target> for SbomTarget {
+    fn from(target: &Target) -> Self {
+        SbomTarget {
+            kind: target.kind().clone(),
+            crate_type: target.kind().rustc_crate_types().first().cloned(),
+            name: target.name().to_string(),
+            edition: target.edition().to_string(),
+        }
+    }
+}
+
+#[derive(Serialize, Clone)]
+struct SbomRustc {
```

Comment:
How do we want to represent a local build rustc?

Reply:
Yes. It will be the same. `rustc.version` is `semver::Version`. However, we'll lose the information that rustc was built from a source tarball, not exactly from official distribution. Is this something SBOM want to track?

```
rustc 1.77.0 (aedd173a2 2024-03-17) (built from a source tarball)
binary: rustc
commit-hash: aedd173a2c086e558c2b66d3743b344f977621a7
commit-date: 2024-03-17
host: x86_64-unknown-linux-gnu
release: 1.77.0
LLVM version: 17.0.6
```

Reply:
I think this information is tracked via [`Rustc.verbose_version`](https://github.com/rust-lang/cargo/blob/master/src/cargo/util/rustc.rs#L27).

Reply:
The verbatim text is present in the SBOM output now.